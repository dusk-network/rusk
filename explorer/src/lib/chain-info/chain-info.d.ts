type BlockHeader = {
  date: Date;
  feespaid: number;
  hash: string;
  height: number;
  nextblockhash: string;
  prevblockhash: string;
  reward: number;
  seed: string;
  statehash: string;
};

type BlockTransactions = {
  data: Transaction[];
  stats: {
    averageGasPrice: number;
    gasLimit: number;
    gasUsed: number;
  };
};

type Block = {
  header: BlockHeader;
  transactions: BlockTransactions;
};

type ChainInfo = {
  blocks: Block[];
  transactions: Transaction[];
};

type SearchResult = {
  id: string;
  type: "account" | "block" | "transaction";
};

type Transaction = {
  amount: number | undefined;
  blobHashes: string[];
  from: string | undefined;
  to: string | undefined;
  blockhash: string;
  blockheight: number;
  date: Date;
  feepaid: number;
  gaslimit: number;
  gasprice: number;
  gasspent: number;
  memo: string;
  method: string;
  success: boolean;
  txerror: string;
  txid: string;
  txtype: string;
  payload: string;
  nonce: number | undefined;
};
