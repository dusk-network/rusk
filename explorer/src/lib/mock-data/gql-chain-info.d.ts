type GQLBlock = {
  fees: number;
  gasSpent: number;
  header: {
    gasLimit: number;
    hash: string;
    height: number;
    nextBlockHash?: string;
    prevBlockHash: string;
    seed: string;
    stateHash: string;
    timestamp: number;
  };
  reward: number;
  transactions: GQLTransaction[];
};

type GQLCallData = {
  contractId: string;
  fnName: string;
};

type GQLChainInfo = {
  blocks: GQLBlock[];
  transactions: GQLTransaction[];
};

type GQLSearchResult = {
  block?: {
    header: {
      hash: string;
    };
  } | null;
  tx?: {
    id: string;
  } | null;
  account?: {
    id: string;
  } | null;
};

type GQLTransaction = {
  blockHash: string;
  blockHeight: number;
  blockTimestamp: number;
  err: string | null;
  gasSpent: number;
  id: string;
  tx: {
    blobHashes: string[] | null;
    callData: GQLCallData | null;
    gasLimit: number;
    gasPrice: number;
    id: string;
    isDeploy: boolean;
    memo: string;
    txType: string;
    json?: string;
  };
};
