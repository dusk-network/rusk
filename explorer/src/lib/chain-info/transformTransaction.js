import { unixTsToDate } from "$lib/dusk/date";

/** @type {(v: GQLTransaction) => Transaction} */
const transformTransaction = (tx) => ({
  blockhash: tx.blockHash,
  blockheight: tx.blockHeight,
  date: unixTsToDate(tx.blockTimestamp),
  feepaid: tx.gasSpent * tx.tx.gasPrice,
  from: undefined,
  gaslimit: tx.tx.gasLimit,
  gasprice: tx.tx.gasPrice,
  gasspent: tx.gasSpent,
  memo: tx.tx.memo ?? "",
  method: tx.tx.isDeploy ? "deploy" : (tx.tx.callData?.fnName ?? "transfer"),
  success: tx.err === null,
  to: undefined,
  txerror: tx.err ?? "",
  txid: tx.id,
  txtype: tx.tx.txType,
});

export default transformTransaction;
