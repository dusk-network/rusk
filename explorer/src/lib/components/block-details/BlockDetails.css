.block-details__list {
  display: grid;
  grid-template-columns: minmax(0, 1fr);
  font-size: 0.875rem;
  column-gap: 1.875rem;
  padding: 0 1.125rem;
}

.block-details__list-anchor {
  display: flex;
}

.block-details__gas-used.dusk-progress-bar {
  width: 12.5rem;
}

.block-details__hash {
  font-family: var(--mono-font-family);
  display: flex;
  gap: 0.625rem;
  align-items: center;
}

.block-details__payload {
  margin-top: 0.625rem;
  padding: 0.625rem 0.875rem;
  line-height: 1.3125rem;
  width: 100%;
  overflow-x: auto;
}

.block-details__payload-switch {
  position: absolute;
  right: -3.8125rem;
}

.block-details__switch-wrapper {
  display: flex;
  align-items: center;
  flex: 1 100%;
  justify-content: space-between;
}

.block-details__list .details-list__term:last-of-type {
  align-items: start;
  margin-bottom: 0.625rem;
  /*
    3.8125rem is the sum of the switch width plus the gap
    between the switch and the dt element
  */
  width: calc(100% - 3.8125rem);
}

@media (min-width: 640px) {
  .block-details__header-heading {
    text-transform: none;
  }

  .block-details__list {
    display: inline-grid;
    grid-template-columns: minmax(0, auto) minmax(0, 3fr);
  }

  .block-details__state-hash {
    padding-top: 0rem;
  }

  .block-details__payload {
    margin-top: 0rem;
  }

  .block-details__payload-switch {
    position: relative;
    right: 0rem;
  }

  .block-details__list .details-list__term:last-of-type {
    align-items: start;
    margin-bottom: 0;
    width: 100%;
  }
}
