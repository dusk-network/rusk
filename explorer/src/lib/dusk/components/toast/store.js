import { get, writable } from "svelte/store";

/**
 * @type {import('svelte/store').Writable<ToastItem[]>}
 * @description Stores each toast as an object in the array
 */
export const toastList = writable([]);

/**
 * @type {import('svelte/store').Writable<Number>}
 * @description Stores the timer fused by all toasts
 */
export const toastTimer = writable(0);

/**
 * @param {string} message
 * @param {string} icon
 * @param {TooltipType} type
 */

function addToast(type, message, icon) {
  const id = `dusk-toast-${Math.random().toString(36)}`;

  toastList.update((store) => {
    return [
      ...store,
      {
        icon: icon,
        id: id,
        message: message,
        type: type,
      },
    ];
  });

  const timeoutID = window.setTimeout(() => {
    deleteToast(id, timeoutID);
  }, get(toastTimer));
}

/**
 * @param {string} id
 * @param {number} timeout
 */
function deleteToast(id, timeout) {
  window.clearTimeout(timeout);

  // Deletes toast from store queue
  toastList.update((store) => {
    return store.filter((toast) => toast.id !== id);
  });
}

export const toast = addToast;
