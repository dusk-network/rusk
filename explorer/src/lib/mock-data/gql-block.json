{
  "block": {
    "fees": 580718,
    "gasSpent": 580718,
    "header": {
      "gasLimit": 5000000000,
      "hash": "bd5c99bb720b03500e89f103fe66113ba62f2e124ed9651563f38fd15977719f",
      "height": 495868,
      "nextBlockHash": "6011556208a85e6001bd01ccbf936486b91318a7f6cbcf7ab810adf6fae34204",
      "prevBlockHash": "07b74b35c2c7cf8f41426cd0870bafa1a2c7adee3fdd876643548096186fc4cb",
      "seed": "af15447e3a004a79d4ae8b084f7b76b78d95880bb63e1cfa79250a310731f52e6d84ee42a5d6fc2cb99c5b1f489761f6",
      "stateHash": "20bb0a677b93f084afadfd34bec3ac3feee33a020b81d9549afa2268e8543acb",
      "timestamp": 1713298022
    },
    "reward": 16000000000,
    "transactions": [
      {
        "blockHash": "bd5c99bb720b03500e89f103fe66113ba62f2e124ed9651563f38fd15977719f",
        "blockHeight": 495868,
        "blockTimestamp": 1713298022,
        "err": null,
        "gasSpent": 290866,
        "id": "3a3f6f90a1012ae751b4448bcb8e98def0ba2b18170239bd69fcf8e2e37f0602",
        "tx": {
          "blobHashes": null,
          "callData": null,
          "isDeploy": false,
          "gasLimit": 500000000,
          "gasPrice": 1,
          "id": "3a3f6f90a1012ae751b4448bcb8e98def0ba2b18170239bd69fcf8e2e37f0602",
          "memo": "",
          "txType": "Moonlight"
        }
      },
      {
        "blockHash": "bd5c99bb720b03500e89f103fe66113ba62f2e124ed9651563f38fd15977719f",
        "blockHeight": 495868,
        "blockTimestamp": 1713298022,
        "err": null,
        "gasSpent": 289852,
        "id": "07bfabea1d94c16f2dc3697fa642f6cecea6e81bf76b9644efbb6e2723b76d00",
        "tx": {
          "blobHashes": null,
          "callData": null,
          "gasLimit": 500000000,
          "gasPrice": 1,
          "id": "07bfabea1d94c16f2dc3697fa642f6cecea6e81bf76b9644efbb6e2723b76d00",
          "isDeploy": false,
          "memo": "some notes about the transaction",
          "txType": "Phoenix"
        }
      }
    ]
  }
}
