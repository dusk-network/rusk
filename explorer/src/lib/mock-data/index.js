export { default as apiMarketData } from "./api-market-data.json";
export { default as apiNodeInfo } from "./api-node-info.json";
export { default as apiNodeLocations } from "./api-node-locations.json";
export { default as apiStats } from "./api-stats.json";
export { default as gqlBlock } from "./gql-block.json";
export { default as gqlBlockDetails } from "./gql-block-details.json";
export { default as gqlBlocks } from "./gql-blocks.json";
export { default as gqlLatestChainInfo } from "./gql-latest-chain-info.json";
export { default as gqlSearchPossibleResults } from "./gql-search-possible-results.json";
export { default as gqlTransaction } from "./gql-transaction.json";
export { default as gqlTransactions } from "./gql-transactions.json";
export { default as hostProvisioners } from "./host-provisioners.json";
export { default as nodeLocationsCount } from "./node-locations.count.json";

export { default as accounts } from "./accounts.json";
export { default as tokens } from "./tokens.json";
export { default as gqlTokenTransactions } from "./gql-token-transactions.json";
