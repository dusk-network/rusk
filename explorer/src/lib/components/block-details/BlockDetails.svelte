<svelte:options immutable={true} />

<script>
  import { mdiArrowLeft, mdiArrowRight } from "@mdi/js";
  import { AppAnchorButton, DataCard, ListItem } from "$lib/components";
  import {
    Card,
    CopyButton,
    ProgressBar,
    RelativeTime,
    Switch,
  } from "$lib/dusk/components";
  import { luxToDusk } from "$lib/dusk/currency";
  import { createValueFormatter } from "$lib/dusk/value";
  import {
    calculateAdaptiveCharCount,
    makeClassName,
    middleEllipsis,
  } from "$lib/dusk/string";
  import { onMount } from "svelte";
  import "./BlockDetails.css";

  /** @type {string | Undefined} */
  export let className = undefined;

  /** @type {Block} */
  export let data;

  /** @type {Error | null}*/
  export let error;

  /** @type {Boolean} */
  export let loading;

  /** @type {String | null} */
  export let payload;

  const formatter = createValueFormatter("en");

  /** @type {Number} */
  let screenWidth = window.innerWidth;

  /** @type {Boolean} */
  let isPayloadToggled = false;

  $: classes = makeClassName(["block-details", className]);

  onMount(() => {
    const resizeObserver = new ResizeObserver((entries) => {
      const entry = entries[0];

      screenWidth = entry.contentRect.width;
    });

    resizeObserver.observe(document.body);

    return () => resizeObserver.disconnect();
  });
</script>

<DataCard
  on:retry
  {data}
  {error}
  {loading}
  className={classes}
  title="Block Details"
>
  <dl class="block-details__list">
    <!-- HEIGHT -->
    <ListItem
      tooltipText="The height of the block indicates the length of the block chain and is increased with each additional block"
    >
      <svelte:fragment slot="term">height</svelte:fragment>
      <svelte:fragment slot="definition">
        <AppAnchorButton
          className="block-details__list-anchor"
          href="/blocks/block?id={data.header.prevblockhash}"
          icon={{ path: mdiArrowLeft }}
          disabled={!data.header.prevblockhash || data.header.height === 0}
        />
        {formatter(data.header.height)}
        <AppAnchorButton
          className="block-details__list-anchor"
          href="/blocks/block?id={data.header.nextblockhash}"
          icon={{ path: mdiArrowRight }}
          disabled={!data.header.nextblockhash}
        />
      </svelte:fragment>
    </ListItem>

    <!-- BLOCK HASH -->
    <ListItem tooltipText="The hash for the header of the block">
      <svelte:fragment slot="term">hash</svelte:fragment>
      <svelte:fragment slot="definition"
        >{middleEllipsis(
          data.header.hash,
          calculateAdaptiveCharCount(screenWidth, 320, 1920, 14, 66)
        )}
        <CopyButton rawValue={data.header.hash} name="Block hash" />
      </svelte:fragment>
    </ListItem>

    <!-- TIMESTAMP -->
    <ListItem tooltipText="The date and time the block was created">
      <svelte:fragment slot="term">timestamp</svelte:fragment>
      <RelativeTime
        autoRefresh={true}
        className="block-details__list-timestamp"
        date={data.header.date}
        slot="definition"
        ><svelte:fragment let:relativeTime
          >{`${data.header.date.toUTCString()} (${relativeTime})`}</svelte:fragment
        ></RelativeTime
      >
    </ListItem>

    <!-- TRANSACTIONS -->
    <ListItem tooltipText="The number of transactions included in the block">
      <svelte:fragment slot="term">transactions</svelte:fragment>
      <svelte:fragment slot="definition"
        >{formatter(data.transactions.data.length)}</svelte:fragment
      >
    </ListItem>

    <!-- BLOCK FEES PAID -->
    <ListItem
      tooltipText="The total fees paid for the transactions in the block"
    >
      <svelte:fragment slot="term">block fees paid</svelte:fragment>
      <svelte:fragment slot="definition"
        >{formatter(luxToDusk(data.header.feespaid))} DUSK</svelte:fragment
      >
    </ListItem>

    <!-- BLOCK REWARD -->
    <ListItem
      tooltipText="The reward allocated to the block generator can fluctuate between 70-80% of the total reward and the flexible 10% can end up being burnt if there's no unanimous vote on a block"
    >
      <svelte:fragment slot="term">block reward</svelte:fragment>
      <svelte:fragment slot="definition"
        >{formatter(luxToDusk(data.header.reward))} DUSK</svelte:fragment
      >
    </ListItem>

    <!-- BLOCK GAS LIMIT -->
    <ListItem tooltipText="The block gas limit">
      <svelte:fragment slot="term">block gas limit</svelte:fragment>
      <svelte:fragment slot="definition"
        >{formatter(data.transactions.stats.gasLimit)}</svelte:fragment
      >
    </ListItem>

    <!-- GAS USED -->
    <ListItem tooltipText="The amount of gas used generating the block">
      <svelte:fragment slot="term">gas used</svelte:fragment>
      <svelte:fragment slot="definition">
        {formatter(data.transactions.stats.gasUsed)}

        <ProgressBar
          currentPercentage={data.transactions.stats.gasLimit === 0
            ? 0
            : (data.transactions.stats.gasUsed /
                data.transactions.stats.gasLimit) *
              100}
          className="block-details__gas-used"
          ariaLabel="Gas Used"
        />
      </svelte:fragment>
    </ListItem>

    <!-- AVERAGE GAS PRICE -->
    <ListItem
      tooltipText="The average gas price for the transactions within the block"
    >
      <svelte:fragment slot="term">average gas price</svelte:fragment>
      <svelte:fragment slot="definition"
        >{formatter(data.transactions.stats.averageGasPrice)}</svelte:fragment
      >
    </ListItem>

    <!-- STATE ROOT HASH -->
    <ListItem tooltipText="The state root hash">
      <svelte:fragment slot="term">state root hash</svelte:fragment>
      <span class="block-details__state-hash" slot="definition"
        >{middleEllipsis(
          data.header.statehash,
          calculateAdaptiveCharCount(screenWidth, 320, 1920, 14, 66)
        )}</span
      >
    </ListItem>

    <!-- HEADER -->
    <ListItem tooltipText="The block header information">
      <svelte:fragment slot="term">
        <div class="block-details__switch-wrapper">
          header

          <Switch
            className="block-details__payload-switch"
            onSurface={true}
            bind:value={isPayloadToggled}
          />
        </div>
      </svelte:fragment>

      <svelte:fragment slot="definition">
        {#if isPayloadToggled}
          <Card onSurface={true} className="block-details__payload">
            <pre>{payload
                ? JSON.stringify(JSON.parse(payload), null, 2)
                : "---"}</pre>
          </Card>
        {/if}
      </svelte:fragment>
    </ListItem>
  </dl>
</DataCard>
