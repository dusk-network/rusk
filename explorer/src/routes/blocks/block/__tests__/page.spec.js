import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, render } from "@testing-library/svelte";
import { get } from "svelte/store";

import { duskAPI } from "$lib/services";
import { transformBlock } from "$lib/chain-info";
import { apiBlock } from "$lib/mock-data";
import { appStore } from "$lib/stores";

import BlockDetails from "../+page.svelte";

global.ResizeObserver = vi.fn().mockImplementation(() => ({
  disconnect: vi.fn(),
  observe: vi.fn(),
  unobserve: vi.fn(),
}));

describe("Block Details", () => {
  vi.useFakeTimers();

  const { fetchInterval } = get(appStore);
  const getBlockSpy = vi
    .spyOn(duskAPI, "getBlock")
    .mockResolvedValue(transformBlock(apiBlock.data.blocks[0]));

  afterEach(() => {
    cleanup();
    getBlockSpy.mockClear();
  });

  afterAll(() => {
    vi.useRealTimers();
    getBlockSpy.mockRestore();
  });

  it("should render the Block Details page, start polling the block data and stop the polling when unmounted", async () => {
    const { container, unmount } = render(BlockDetails);

    expect(container.firstChild).toMatchSnapshot();

    expect(getBlockSpy).toHaveBeenCalledTimes(1);

    await vi.advanceTimersByTimeAsync(fetchInterval);

    expect(getBlockSpy).toHaveBeenCalledTimes(2);

    await vi.advanceTimersByTimeAsync(fetchInterval);

    expect(getBlockSpy).toHaveBeenCalledTimes(3);

    unmount();

    await vi.advanceTimersByTimeAsync(fetchInterval * 10);

    expect(getBlockSpy).toHaveBeenCalledTimes(3);
  });
});
