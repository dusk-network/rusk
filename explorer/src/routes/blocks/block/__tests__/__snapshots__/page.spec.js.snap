// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Block Details > should render the Block Details page and query the necessary info 1`] = `
<section
  class="block svelte-qy5n2j"
>
  <div
    class="block__details"
  >
    <div
      class="dusk-card data-card block-details"
    >
      <header
        class="data-card__header"
        slot="header"
      >
        <h1
          class="data-card__header-title"
        >
          Block Details
        </h1>
         
      </header>
       
      <div
        aria-label="Loading"
        class="dusk-progress-bar data-card__progress-bar"
        role="progressbar"
      >
        <div
          class="dusk-progress-bar__filler dusk-progress-bar__filler--undetermined"
        />
      </div>
      
       
    </div>
    
  </div>
   
  <div
    class="block__transactions"
  >
    <div
      class="dusk-card data-card latest-transactions-card"
    >
      <header
        class="data-card__header"
        slot="header"
      >
        <h1
          class="data-card__header-title"
        >
          Transactions
        </h1>
         
      </header>
       
      <div
        aria-label="Loading"
        class="dusk-progress-bar data-card__progress-bar"
        role="progressbar"
      >
        <div
          class="dusk-progress-bar__filler dusk-progress-bar__filler--undetermined"
        />
      </div>
      
       
    </div>
    
    
  </div>
</section>
`;

exports[`Block Details > should render the Block Details page and query the necessary info 2`] = `
<section
  class="block svelte-qy5n2j"
>
  <div
    class="block__details"
  >
    <div
      class="dusk-card data-card block-details"
    >
      <header
        class="data-card__header"
        slot="header"
      >
        <h1
          class="data-card__header-title"
        >
          Block Details
        </h1>
         
      </header>
       
      <div
        class="data-card__content"
      >
        <dl
          class="block-details__list"
        >
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The height of the block indicates the length of the block chain and is increased with each additional block"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            height
          </dt>
           
          <dd
            class="details-list__definition"
          >
            <a
              aria-disabled="false"
              class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--primary dusk-anchor-button--size--normal dusk-icon-button block-details__list-anchor"
              href="/some-base-path/blocks/block?id=07b74b35c2c7cf8f41426cd0870bafa1a2c7adee3fdd876643548096186fc4cb"
            >
              <svg
                class="dusk-icon dusk-icon--size--normal dusk-anchor-button__icon"
                role="graphics-symbol"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
                />
              </svg>
              
               
              
              
            </a>
             
            495,868
             
            <a
              aria-disabled="false"
              class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--primary dusk-anchor-button--size--normal dusk-icon-button block-details__list-anchor"
              href="/some-base-path/blocks/block?id=6011556208a85e6001bd01ccbf936486b91318a7f6cbcf7ab810adf6fae34204"
            >
              <svg
                class="dusk-icon dusk-icon--size--normal dusk-anchor-button__icon"
                role="graphics-symbol"
                viewBox="0 0 24 24"
              >
                <path
                  d="M4,11V13H16L10.5,18.5L11.92,19.92L19.84,12L11.92,4.08L10.5,5.5L16,11H4Z"
                />
              </svg>
              
               
              
              
            </a>
             
          </dd>
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The hash for the header of the block"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            hash
          </dt>
           
          <dd
            class="details-list__definition"
          >
            bd5c99bb720b03500e89f103fe66113ba62f2e124ed9651563f38fd15977719f
          </dd>
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The date and time the block was created"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            timestamp
          </dt>
           
          <dd
            class="details-list__definition"
          >
            <time
              class="block-details__list-timestamp"
              datetime="2024-04-16T20:07:02.000Z"
              slot="definition"
            >
              Tue, 16 Apr 2024 20:07:02 GMT (last month)
              
               
            </time>
          </dd>
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The number of transactions included in the block"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            transactions
          </dt>
           
          <dd
            class="details-list__definition"
          >
            2
          </dd>
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The total fees paid for the transactions in the block"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            block fees paid
          </dt>
           
          <dd
            class="details-list__definition"
          >
            0.000580718
             DUSK
          </dd>
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The reward allocated to the block generator"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            block reward
          </dt>
           
          <dd
            class="details-list__definition"
          >
            16
             DUSK
          </dd>
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The block gas limit"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            block gas limit
          </dt>
           
          <dd
            class="details-list__definition"
          >
            5,000,000,000
          </dd>
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The amount of gas used generating the block"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            gas used
          </dt>
           
          <dd
            class="details-list__definition"
          >
            580,718
             
            <div
              aria-label="Gas Used"
              class="dusk-progress-bar block-details__gas-used"
              role="progressbar"
            >
              <div
                class="dusk-progress-bar__filler"
                style="width: 0.01161436%"
              />
            </div>
             
          </dd>
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The average fee paid for the transactions within the block"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            average fee paid
          </dt>
           
          <dd
            class="details-list__definition"
          >
            0.000000001
             DUSK
          </dd>
          <dt
            class="details-list__term"
          >
            <svg
              class="dusk-icon dusk-icon--size--normal detail-list__help"
              data-tooltip-id="main-tooltip"
              data-tooltip-place="top"
              data-tooltip-text="The state root hash"
              data-tooltip-type="info"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
            
             
            state root hash
          </dt>
           
          <dd
            class="details-list__definition"
          >
            <span
              class="block-details__state-hash"
              slot="definition"
            >
              20bb0a677b93f084afadfd34bec3ac3feee33a020b81d9549afa2268e8543acb
            </span>
          </dd>
        </dl>
      </div>
      
       
    </div>
    
  </div>
   
  <div
    class="block__transactions"
  >
    <div
      class="dusk-card data-card latest-transactions-card"
    >
      <header
        class="data-card__header"
        slot="header"
      >
        <h1
          class="data-card__header-title"
        >
          Transactions
        </h1>
         
      </header>
       
      <div
        class="data-card__content"
      >
        <div
          class="table-container transactions-table"
        >
          <table
            class="table"
          >
            <thead
              class="table__head"
            >
              <tr
                class="table__row"
              >
                <th
                  class="table__header-cell"
                >
                  ID
                </th>
                
                 
                <th
                  class="table__header-cell"
                >
                  Gas
                </th>
                
                 
                <th
                  class="table__header-cell"
                >
                  Fee (DUSK)
                </th>
                
                 
                <th
                  class="table__header-cell"
                >
                  Status
                </th>
                
                 
                <th
                  class="table__header-cell"
                >
                  Type
                </th>
                
              </tr>
            </thead>
             
            <tbody
              class="table__body"
            >
              
              <tr
                class="table__row"
              >
                <td
                  class="table__data-cell"
                >
                  <a
                    class="dusk-anchor dusk-anchor--on-surface transaction__link"
                    href="/some-base-path/transactions/transaction?id=3a3f6f90a1012ae751b4448bcb8e98def0ba2b18170239bd69fcf8e2e37f0602"
                  >
                    3a3f6f90a1...e2e37f0602
                  </a>
                   
                  <small
                    class="transaction__time"
                  >
                    last month
                  </small>
                </td>
                
                 
                <td
                  class="table__data-cell"
                >
                  <b
                    class="transaction__fee-price-label"
                  >
                    PRICE:
                  </b>
                   
                  1
                  <br />
                   
                  <b
                    class="transaction__fee-limit-label"
                  >
                    LIMIT:
                  </b>
                   
                  500,000,000
                </td>
                
                 
                <td
                  class="table__data-cell"
                >
                  0.000290866
                </td>
                
                 
                <td
                  class="table__data-cell"
                >
                  <span
                    class="dusk-badge dusk-badge--variant-success"
                  >
                    success
                  </span>
                </td>
                
                 
                <td
                  class="table__data-cell"
                >
                  <span
                    class="dusk-badge dusk-badge--variant-neutral transaction__type"
                  >
                    transfer
                  </span>
                </td>
                
                 
              </tr>
              
              <tr
                class="table__row"
              >
                <td
                  class="table__data-cell"
                >
                  <a
                    class="dusk-anchor dusk-anchor--on-surface transaction__link"
                    href="/some-base-path/transactions/transaction?id=07bfabea1d94c16f2dc3697fa642f6cecea6e81bf76b9644efbb6e2723b76d00"
                  >
                    07bfabea1d...2723b76d00
                  </a>
                   
                  <small
                    class="transaction__time"
                  >
                    last month
                  </small>
                </td>
                
                 
                <td
                  class="table__data-cell"
                >
                  <b
                    class="transaction__fee-price-label"
                  >
                    PRICE:
                  </b>
                   
                  1
                  <br />
                   
                  <b
                    class="transaction__fee-limit-label"
                  >
                    LIMIT:
                  </b>
                   
                  500,000,000
                </td>
                
                 
                <td
                  class="table__data-cell"
                >
                  0.000289852
                </td>
                
                 
                <td
                  class="table__data-cell"
                >
                  <span
                    class="dusk-badge dusk-badge--variant-success"
                  >
                    success
                  </span>
                </td>
                
                 
                <td
                  class="table__data-cell"
                >
                  <span
                    class="dusk-badge dusk-badge--variant-neutral transaction__type"
                  >
                    transfer
                  </span>
                </td>
                
                 
              </tr>
              
            </tbody>
          </table>
        </div>
        
      </div>
      
       
    </div>
    
    
  </div>
</section>
`;
